truss-chains/truss_chains/deploy.py

```diff
@@ -338,71 +328,423 @@ def deploy_remotely(
+            oracle_version_id=version_id,
+            oracle_predict_url=url,
+        )
+    return chainlet_data
+
+
+class _LivePatcher:
+    _source: pathlib.Path
+    _entrypoint: Optional[str]
+    _deployed_chain_name: str
+    _remote_provider: b10_remote.BasetenRemote
+    _chainlet_data: Mapping[str, _ChainletData]
+    _watch_filter: Callable[[watchfiles.Change, str], bool]
+    _console: "rich_console.Console"
+    _error_console: "rich_console.Console"
+
+    def __init__(
+        self,
+        source: pathlib.Path,
+        entrypoint: Optional[str],
+        name: Optional[str],
+        remote: Optional[str],
+        console: "rich_console.Console",
+        error_console: "rich_console.Console",
+    ) -> None:
+        self._source = source
+        self._entrypoint = entrypoint
+        self._console = console
+        self._error_console = error_console
+        if not remote:
+            remote = remote_cli.inquire_remote_name(
+                remote_factory.RemoteFactory.get_available_config_names()
+            )
+        self._remote_provider = cast(
+            b10_remote.BasetenRemote,
+            remote_factory.RemoteFactory.create(remote=remote),
         )
+        with framework.import_target(source, entrypoint) as entrypoint_cls:
+            self._deployed_chain_name = name or entrypoint_cls.__name__
+            self._chain_root = _get_chain_root(entrypoint_cls)
+            chainlet_names = set(
+                desc.display_name
+                for desc in _get_ordered_dependencies([entrypoint_cls])
+            )
 
-        chain_service.set_remote_chain_service(
-            chain_deployment_handle.chain_id,
-            chain_deployment_handle.chain_deployment_id,
-            options.remote_provider,
+        chain_id = b10_core.get_chain_id_by_name(
+            self._remote_provider.api, self._deployed_chain_name
+        )
+        if not chain_id:
+            raise definitions.ChainsDeploymentError(
+                f"Chain `{chain_id}` was not found."
+            )
+        self._status_page_url = _chain_status_page_url(
+            self._remote_provider.remote_url, chain_id
+        )
+        chain_deployment = b10_core.get_dev_chain_deployment(
+            self._remote_provider.api, chain_id
         )
+        if chain_deployment is None:
+            raise definitions.ChainsDeploymentError(
+                f"No development deployment was found for Chain `{chain_id}`. "
+                "You cannot live-patch production deployments. Check the Chain's "
+                f"status page for available deployments: {self._status_page_url}."
+            )
+        deployed_chainlets = self._remote_provider.get_chainlets(chain_deployment["id"])
+        non_draft_chainlets = [
+            chainlet["name"]
+            for chainlet in deployed_chainlets
+            if not chainlet["oracle_version"]["is_draft"]
+        ]
+        assert not (
```

Comment:
how does this show up to user?

Reply:
It's an assertion. That means it cannot be triggered by the user providing wrong inputs or using it the wrong way, but only by a system-internal bug.

I follow this rationale to decide between exceptions and assertions:

Assertions should only be used to verify conditions that should be logically impossible to be false (read: sanity checks). These conditions should only be based on inputs generated by your own code. Any checks based on external inputs should use exceptions.

https://softwareengineering.stackexchange.com/questions/15515/when-to-use-assertions-and-when-to-use-exceptions